replica/server.go

```diff
@@ -360,6 +360,74 @@ func (s *Server) Revert(name, created string) error {
 	return nil
 }
 
+// UpdateLUNMap updates the the original LUNmap with any changes which have been
+// encountered after PreloadVolume
+func (s *Server) UpdateLUNMap() error {
+	// With this lock r.volume data structure is copied, and a new slice is
+	// created different than the one in r.volume.location. After this we will
+	// be having 2 LUNMaps, one being filled by the parallel write operations
+	// and the other being filled by preload operation.
+	s.Lock()
+	volume := s.r.volume
+	volume.location = make([]uint16, len(s.r.volume.location))
+	s.Unlock()
+	// LUNmap is populated with the extents after the sync operation is
+	// completed
+	if err := PreloadLunMap(&volume); err != nil {
+		return err
+	}
+	s.Lock()
+	var (
+		holeLength          int64
+		holeOffset          int
+		fileIndx            uint16
+		prevHoleFileIndx    uint16
+		offset              int
+		userCreatedSnapIndx uint16
+	)
+	// userCreatedSnapIndx holds the latest user created snapshot index
+	for i, isUserCreated := range volume.UserCreatedSnap {
+		if isUserCreated {
+			userCreatedSnapIndx = uint16(i)
+		}
+	}
+
+	// While this loop is being executed IOs have been stopped(s.Lock()) which
+	// inturn halts the updates on the original LunMap.
+	// Empty offsets in the original LunMap are filled by corresponding entries
+	// in preloaded lunmap.
+	// If offsets are present in both LunMaps and are different,
+	// hole is punched in the file at that offset contained in Preloaded LunMap.
+	// Sequesnce of holes are being punched at once.
+	for offset, fileIndx = range volume.location {
+		if fileIndx == 0 {
+			continue
+		}
+		if s.r.volume.location[offset] > fileIndx {
+			if prevHoleFileIndx != fileIndx || int64(offset) != int64(holeOffset)+holeLength {
+				if prevHoleFileIndx > userCreatedSnapIndx && shouldCreateHoles() && prevHoleFileIndx != 0 {
+					sendToCreateHole(volume.files[prevHoleFileIndx], int64(holeOffset)*volume.sectorSize, holeLength*volume.sectorSize)
+				}
+				holeLength = 1
+				holeOffset = offset
+				prevHoleFileIndx = fileIndx
+			} else {
+				holeLength++
+			}
+		} else {
+			// No hole drilling over here as that offset is empty
+			s.r.volume.location[offset] = volume.location[offset]
```

Comment:
we need to be copying other variables of diffDisk like UsedBlocks and UsedLogicalBlocks and testcases for them

Reply:
Also, it seems some problem in calculating UsedBlocks of old code.. we are NOT decrementing it when the holes are punched

Reply:
fixed